hyperspace/cosmos/src/provider.rs

```diff
@@ -58,11 +128,118 @@ where
+
+			let tx_results = block_results.txs_results.ok_or_else(|| Err(Error::Custom("empty transaction results".to_string())))?;
+			for tx in tx_results.iter() {
+				for event in tx.events {
+					let ibc_event = ibc_event_try_from_abci_event(&event)?;
+					ibc_events.push(ibc_event);
+				}
+			}
+		}
+
+		let update_header = CosmosClient::msg_update_client_header(self, cs.latest_height).await?;
+		let update_client_header = {
+			let msg = MsgUpdateAnyClient::<LocalClientTypes> {
+				client_id: self.client_id(),
+				client_message: AnyClientMessage::Tendermint(ClientMessage::Header(update_header.0)),
+				signer: counterparty.account_id(),
+			};
+			let value = msg.encode_vec();
+			Any { value, type_url: msg.type_url() }
+		};
+		Ok((update_client_header, ibc_events, update_header.1))
 	}
 
 	async fn ibc_events(&self) -> Pin<Box<dyn Stream<Item = IbcEvent>>> {
-		todo!()
+		let (ws_client, ws_driver) = WebSocketClient::new(self.websocket_url.clone())
+			.await
+			.map_err(|e| Error::from(format!("Web Socket Client Error {:?}", e)))
+			.unwrap();
+		let driver_handle = std::thread::spawn(|| ws_driver.run());
+
+		// ----
+		let query_all = vec![
+			Query::from(EventType::NewBlock),
+			Query::eq("message.module", "ibc_client"),
+			Query::eq("message.module", "ibc_connection"),
+			Query::eq("message.module", "ibc_channel"),
+		];
+
+		let mut subscriptions = vec![];
+		for query in &query_all {
+			let subscription = ws_client
+				.subscribe(query.clone())
+				.await
+				.map_err(|e| Error::from(format!("Web Socket Client Error {:?}", e)));
+			subscriptions.push(subscription);
+		}
+
+		let all_subscribtions = Box::new(select_all(subscriptions));
+		// Collect IBC events from each RPC event
+		let events = all_subscribtions
+			.map_ok(move |event| {
+				let mut events: Vec<IbcEvent> = vec![];
+				let Event { data, events, query } = event;
+				match data {
+					EventData::NewBlock { block, .. }
+					if query == Query::from(EventType::NewBlock).to_string() =>
+						{
+							events.push(NewBlock::new(height).into());
+							// events_with_height.append(&mut extract_block_events(height, &events));
+						},
+					EventData::Tx { tx_result } => {
+						for abci_event in &tx_result.result.events {
+							if let Ok(ibc_event) = ibc_event_try_from_abci_event(abci_event) {
+								if query == Query::eq("message.module", "ibc_client").to_string()
+									&& event_is_type_client(&ibc_event)
+								{
+									events.push(ibc_event);
+								} else if query
+									== Query::eq("message.module", "ibc_connection").to_string()
+									&& event_is_type_connection(&ibc_event)
+								{
+									events.push(ibc_event);
+								} else if query
+									== Query::eq("message.module", "ibc_channel").to_string()
+									&& event_is_type_channel(&ibc_event)
+								{
+									events.push(ibc_event);
+								}
+							}
+						}
+					},
+					_ => {},
+				}
+				stream::iter(events).map(Ok)
```

Comment:
what's this piece trying to achieve?

Reply:
This code was gotten from this PR https://github.com/ComposableFi/centauri/pull/103/files#diff-b0380e22a93a5d0cd77b29c49b7bfd1b21aadc9cfc193ecb5d6edc0135e39178R5. It will be refactored on the next PR when we add conversion helper methods to the existing event types.